operator/pkg/ingress/ingress_class.go

```diff
@@ -4,172 +4,181 @@
+	ingressClasses        resource.Resource[*slim_networkingv1.IngressClass]
+	queue                 workqueue.RateLimitingInterface
+	events                <-chan resource.Event[*slim_networkingv1.IngressClass]
+}
+
+func newIngressClassManager(
+	ctx context.Context,
+	queue workqueue.RateLimitingInterface,
+	ingressClasses resource.Resource[*slim_networkingv1.IngressClass],
+) (*ingressClassManager, error) {
 	manager := &ingressClassManager{
-		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		ingressQueue: ingressQueue,
-		maxRetries:   maxRetries,
+		isDefaultIngressClass: atomic.Bool{},
+		synced:                atomic.Bool{},
+		ingressClasses:        ingressClasses,
+		events:                ingressClasses.Events(context.Background()),
+		queue:                 queue,
 	}
 
-	manager.store, manager.informer = informer.NewInformer(
-		utils.ListerWatcherFromTyped[*slim_networkingv1.IngressClassList](clientset.Slim().NetworkingV1().IngressClasses()),
-		&slim_networkingv1.IngressClass{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    manager.handleAddIngressClass,
-			UpdateFunc: manager.handleUpdateIngressClass,
-			DeleteFunc: manager.handleDeleteIngressClass,
-		},
-		nil,
-	)
-
-	go manager.informer.Run(wait.NeverStop)
-	if !cache.WaitForCacheSync(wait.NeverStop, manager.informer.HasSynced) {
-		return manager, fmt.Errorf("unable to sync service")
-	}
-	log.WithField("existing-ingressclasses", manager.store.ListKeys()).Debug("ingress classes synced")
+	manager.isDefaultIngressClass.Store(false)
+	manager.synced.Store(false)
+
 	return manager, nil
 }
 
-// Run kicks off the control loop
-func (i *ingressClassManager) Run() {
-	for i.processEvent() {
+func (i *ingressClassManager) IsDefault() bool {
+	return i.isDefaultIngressClass.Load()
+}
+
+func (i *ingressClassManager) processEvents(ctx context.Context, initialSync bool) error {
+	for {
+		select {
+		case event := <-i.events:
+			if event.Kind == resource.Sync {
+				i.handleSyncEvent(event)
+
+				// If we are only supposed to do an initial
+				// sync, then exit when we received a sync event.
+				if initialSync {
+					return nil
+				}
+
+				break
+			}
+
+			if event.Object != nil && event.Object.GetName() != ciliumIngressClassName {
+				event.Done(nil)
+
+				break
+			}
+
+			// !initialSync is passed as the value for enqueue in the below handlers
+			// because we should only pass events up to the ingress controller to
+			// allow it to react to changes in the state of the Cilium IngressClass.
+			// When performing an initial sync of the resource, we just need to
+			// learn if cilium is the default IngressClass and expose this
+			// information to the ingress controller via the IsDefault method.
+
+			switch event.Kind {
+			case resource.Upsert:
+				i.handleUpsertEvent(event, !initialSync)
+			case resource.Delete:
+				i.handleDeleteEvent(event, !initialSync)
+			default:
+				if event.Done != nil {
```

Comment:
i don't think that this `nil` check is necessary & common. The function should be set on all events.

Reply:
There must be some other bug going on then. I agree with you that this isn't common, but if I take this nil check out, I get a panic in my tests. I added a log in the default case of the switch statement to show the event struct and it looks empty:

```
level=info msg=Invoked duration="15.411µs" function="ingress.Test_ingressClassHandleEvent.func11.2 (pkg/ingress/ingress_class_test.go:426)" subsys=hive
level=info msg=Starting subsys=hive
level=info msg="Start hook executed" duration="1.702µs" function="*resource.resource[*v1.IngressClass].Start" subsys=hive
level=info msg="IngressClass resources synced" cilium-is-default=false subsys=ingress-controller
level=info msg=Stopping subsys=hive
level=warning msg="got event in default case" event="{  nil <nil>}" subsys=ingress-controller
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x248bf11]

goroutine 63 [running]:
github.com/cilium/cilium/operator/pkg/ingress.(*ingressClassManager).processEvents(0xc00033de60, {0x3115718, 0xc000571900}, 0x0)
	/home/rydrew/Documents/cilium/operator/pkg/ingress/ingress_class.go:93 +0x3d1
github.com/cilium/cilium/operator/pkg/ingress.(*ingressClassManager).Run(0x0?, {0x3115718, 0xc000571900})
	/home/rydrew/Documents/cilium/operator/pkg/ingress/ingress_class.go:126 +0x72
created by github.com/cilium/cilium/operator/pkg/ingress.Test_ingressClassHandleEvent.func11 in goroutine 53
	/home/rydrew/Documents/cilium/operator/pkg/ingress/ingress_class_test.go:446 +0x81e
FAIL	github.com/cilium/cilium/operator/pkg/ingress	0.127s
FAIL
```

Reply:
I think this is happening after `hive.Stop` is called: the `ingressClassEvents` channel is closed and the `select` in `(i *ingressClassManager) processEvents` might unblock on the first case with a bogus event. This is due to the fact that, even if the `ctx` is cancelled and the other select case can run too, `select` chooses pseudo-randomly which path to take.

With this:

```diff
@@ -128,9 +128,12 @@ func (i *ingressClassManager) processEvents(ctx context.Context) error {
 
        for {
                select {
-               case event := <-i.ingressClassEvents:
-                       err = nil
+               case event, ok := <-i.ingressClassEvents:
+                       if !ok {
+                               return nil
+                       }
 
+                       err = nil
                        switch event.Kind {
                        case resource.Sync:
                                err = i.handleSyncEvent()
@@ -140,9 +143,7 @@ func (i *ingressClassManager) processEvents(ctx context.Context) error {
                                err = i.handleDeleteEvent(event)
                        }
 
-                       if event.Done != nil {
-                               event.Done(err)
-                       }
+                       event.Done(err)
                case <-ctx.Done():
                        return ctx.Err()
                }
```

the tests run without any panic

Reply:
:point_up: yep, we must take the closed channel into account

Reply:
Ah right ok, that makes sense. I'll make the change!